File: 2021/12/16/main.py
class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n == 1:
            return [0]

        ns = [[] for _ in range(n)]
        cn = [0] * n
        for e in edges:
            ns[e[0]].append(e[1])
            ns[e[1]].append(e[0])
            cn[e[0]]+=1
            cn[e[1]]+=1

        ls = []
        for i in range(n):
            if cn[i] == 1:
                ls.append(i)

        while n > 2:
            nls = []
            for leaf in ls:
                n-=1
                for nd in ns[leaf]:
                    cn[nd]-=1
                    if cn[nd] == 1:
                        nls.append(nd)
            ls = nls

        res = []
        for v in ls:
            res.append(v)

        return res

File: 2021/12/16/main.go
package main

func findMinHeightTrees(n int, edges [][]int) []int {
	if n == 1 {
		return []int{0}
	}

	ns := make([][]int, n)
	for i := 0; i < n; i++ {
		ns[i] = make([]int, 0, 5)
	}

	cn := make([]int, n)
	for _, e := range edges {
		ns[e[0]] = append(ns[e[0]], e[1])
		ns[e[1]] = append(ns[e[1]], e[0])
		cn[e[0]]++
		cn[e[1]]++
	}

	var ls = make([]int, 0, n)
	for i := 0; i < n; i++ {
		if cn[i] == 1 {
			ls = append(ls, i)
		}
	}

	var nd, leaf int

	for n > 2 {
		nls := make([]int, 0, len(ls))
		for _, leaf = range ls {
			n--
			for _, nd = range ns[leaf] {
				cn[nd]--
				if cn[nd] == 1 {
					nls = append(nls, nd)
				}
			}
		}
		ls = nls
	}

	var res = make([]int, 0, len(ls))
	for _, v := range ls {
		res = append(res, v)
	}

	return res
}
